pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -0,0 +1,265 @@
+func (c *Controller) processNextWorkItem() bool {
+	pvcKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(pvcKey)
+
+	pvcNamespace, pvcName, err := cache.SplitMetaNamespaceKey(pvcKey.(string))
+	if err == nil {
+		err = c.processPVC(pvcNamespace, pvcName)
+		if err == nil {
+			c.queue.Forget(pvcKey)
+			return true
+		}
+	}
+
+	utilruntime.HandleError(fmt.Errorf("PVC %v failed with : %v", pvcKey, err))
+	c.queue.AddRateLimited(pvcKey)
+
+	return true
+}
+
+func (c *Controller) processPVC(pvcNamespace, pvcName string) error {
+	glog.V(4).Infof("Processing PVC %s/%s", pvcNamespace, pvcName)
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished processing PVC %s/%s (%v)", pvcNamespace, pvcName, time.Now().Sub(startTime))
+	}()
+
+	pvc, err := c.pvcLister.PersistentVolumeClaims(pvcNamespace).Get(pvcName)
+	if apierrs.IsNotFound(err) {
+		glog.V(4).Infof("PVC %s/%s not found, ignoring", pvcNamespace, pvcName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if volumeutil.IsPVCBeingDeleted(pvc) && volumeutil.IsProtectionFinalizerPresent(pvc) {
+		// PVC should be deleted. Check if it's used and remove finalizer if
+		// it's not.
+		isUsed, err := c.isBeingUsed(pvc)
+		if err != nil {
+			return err
+		}
+		if !isUsed {
+			return c.removeFinalizer(pvc)
+		}
+	}
+
+	if !volumeutil.IsPVCBeingDeleted(pvc) && !volumeutil.IsProtectionFinalizerPresent(pvc) {
+		// PVC is not being deleted -> it should have the finalizer.
+		// The finalizer should be added by admission plugin, this is just
+		// a safeguard.
+		return c.addFinalizer(pvc)
+	}
+	return nil
+}
+
+func (c *Controller) addFinalizer(pvc *v1.PersistentVolumeClaim) error {
+	claimClone := pvc.DeepCopy()
+	volumeutil.AddProtectionFinalizer(claimClone)
+	_, err := c.client.CoreV1().PersistentVolumeClaims(claimClone.Namespace).Update(claimClone)
+	if err != nil {
+		glog.V(3).Infof("Error adding protection finalizer to PVC %s/%s: %v", pvc.Namespace, pvc.Name)
+		return err
+	}
+	glog.V(3).Infof("Added protection finalizer to PVC %s/%s", pvc.Namespace, pvc.Name)
+	return nil
+}
+
+func (c *Controller) removeFinalizer(pvc *v1.PersistentVolumeClaim) error {
+	claimClone := pvc.DeepCopy()
+	volumeutil.RemoveProtectionFinalizer(claimClone)
+	_, err := c.client.CoreV1().PersistentVolumeClaims(claimClone.Namespace).Update(claimClone)
+	if err != nil {
+		glog.V(3).Infof("Error removing protection finalizer from PVC %s/%s: %v", pvc.Namespace, pvc.Name, err)
+		return err
+	}
+	glog.V(3).Infof("Removed protection finalizer from PVC %s/%s", pvc.Namespace, pvc.Name)
+	return nil
+}
+
+func (c *Controller) isBeingUsed(pvc *v1.PersistentVolumeClaim) (bool, error) {
+	pods, err := c.podLister.Pods(pvc.Namespace).List(labels.Everything())
+	if err != nil {
+		return false, err
+	}
+	for _, pod := range pods {
+		if pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed && pod.Status.Phase != v1.PodUnknown {
```

Comment:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md#pvc-finalizing-controller references referring to a node (pod.Spec.NodeName).. I don't see that check here

Reply:
also, tweak this to `continue` early, rather than nesting

```
if pod.Spec.NodeName == "" {
  continue
}
if pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed {
  continue
}
...
```

Reply:
need ack from @kubernetes/sig-storage-pr-reviews @kubernetes/sig-node-pr-reviews on which pod phase implies the kubelet is done unmounting the volume - I'm not sure Unknown is safe

Reply:
When the pod is in failed phase, then it is done unmounting.

I guess there is no concern about potentially corrupting the data if we delete it while it's being unmounted, since we're going to end up deleting it in the end anyway...

Reply:
I would wait for positive confirmation the pod is terminated (failed/succeeded phase, sounds like?)

We don't know the retain policy on the PV, I wouldn't assume data corruption is acceptable

Reply:
It looks like the PR got updated to use an existing helper method for checking pod termination, which is failed or succeeded, so we should be fine now.

Only if the PV retain policy is Delete, then we would've run the risk of deleting the volume while it was unmounting.

Reply:
> Only if the PV retain policy is Delete, then we would've run the risk of deleting the volume while it was unmounting.

I thought there were also corruption cases possible if a PV got reused by another PVC, but either way, sounds like we're waiting for the right thing now

Reply:
I am afraid there is no state change of anything when a PVC gets unmounted and detached.

We have three retention policies:

- delete: it does not matter if the PV gets deleted  before it's unmounted & detached. Note that cloud providers don't allow that anyway (and PV controller re-tries deletion until it succeeds).

- recycle: supported only on NFS. The last pod that used the PV is stopped before PVC is deleted (endured by this controller). So it's fine to start recycling. NFS does not mind being mounted several times.

- retain: the PV will keep its data. It does not matter if PVC was deleted too early, the PV will be unmounted & detached in the usual way. Only admin can bind retained PV to another PVC.

If we want to keep PVs & PVCs from being deleted while they're attached or mounted, then kubelet and A/D controller need to add their finalizers. Currently they can unmount & detach volumes even if the corresponding PVC & PV was deleted. There are lot of corner cases around that and we could delete lot of tricky code if the PV & PVC were available until they're unmounted & detached, but we have to support inline volumes in pods and they can be deleted at any time... I wonder, would be a finalizer on a pod allowed? It seems quite dangerous to me.